xds/internal/xdsclient/authority.go

```diff
@@ -306,365 +337,295 @@ func (a *authority) updateResourceStateAndScheduleCallbacks(rType xdsresource.Ty
-// resource type and resource names. Even though this method does not directly
-// access the resource cache, it is important that `resourcesMu` be held when
-// calling this method to ensure that a consistent snapshot of resource names is
-// being requested.
-func (a *authority) sendDiscoveryRequestLocked(rType xdsresource.Type, resources map[string]*resourceState) {
-	resourcesToRequest := make([]string, len(resources))
-	i := 0
-	for name := range resources {
-		resourcesToRequest[i] = name
-		i++
+// xdsChannelToUseLocked returns the xdsChannel to use for communicating with
+// the management server. If an active channel is available, it returns that.
+// Otherwise, it creates a new channel using the first server configuration in
+// the list of configurations, and returns that.
+//
+// Only executed in the context of a serializer callback.
+func (a *authority) xdsChannelToUseLocked() *xdsChannelWithConfig {
+	if a.activeXDSChannel != nil {
+		return a.activeXDSChannel
 	}
-	a.transport.SendRequest(rType.TypeURL(), resourcesToRequest)
+
+	sc := a.xdsChannelConfigs[0].sc
+	xc, cleanup, err := a.getChannelForADS(sc, a)
+	if err != nil {
+		a.logger.Warningf("Failed to create xDS channel: %v", err)
+		return nil
+	}
+	a.xdsChannelConfigs[0].xc = xc
+	a.xdsChannelConfigs[0].cleanup = cleanup
+	a.activeXDSChannel = a.xdsChannelConfigs[0]
+	return a.activeXDSChannel
 }
 
-func (a *authority) reportLoad() (*load.Store, func()) {
-	return a.transport.ReportLoad()
+// closeXDSChannelsLocked closes all the xDS channels associated with this
+// authority, when there are no more watchers for any resource type.
+//
+// Only executed in the context of a serializer callback.
+func (a *authority) closeXDSChannelsLocked() {
+	for _, xc := range a.xdsChannelConfigs {
+		if xc.cleanup != nil {
+			xc.cleanup()
+			xc.cleanup = nil
+		}
+		xc.xc = nil
+	}
+	a.activeXDSChannel = nil
 }
 
 func (a *authority) dumpResources() []*v3statuspb.ClientConfig_GenericXdsConfig {
-	a.resourcesMu.Lock()
-	defer a.resourcesMu.Unlock()
-
 	var ret []*v3statuspb.ClientConfig_GenericXdsConfig
-	for rType, resourceStates := range a.resources {
-		typeURL := rType.TypeURL()
-		for name, state := range resourceStates {
-			var raw *anypb.Any
-			if state.cache != nil {
-				raw = state.cache.Raw()
-			}
-			config := &v3statuspb.ClientConfig_GenericXdsConfig{
-				TypeUrl:      typeURL,
-				Name:         name,
-				VersionInfo:  state.md.Version,
-				XdsConfig:    raw,
-				LastUpdated:  timestamppb.New(state.md.Timestamp),
-				ClientStatus: serviceStatusToProto(state.md.Status),
-			}
-			if errState := state.md.ErrState; errState != nil {
-				config.ErrorState = &v3adminpb.UpdateFailureState{
-					LastUpdateAttempt: timestamppb.New(errState.Timestamp),
-					Details:           errState.Err.Error(),
-					VersionInfo:       errState.Version,
+	done := make(chan struct{})
+
+	a.xdsClientSerializer.TrySchedule(func(context.Context) {
```

Comment:
Optional: I find this long inline definition of this function as first argument to callback serializer hard to read. Perhaps pull into a local or define as a handle like others?